File: ios/Vocal/Processing/JobModels.swift
import Foundation

struct InputDeclaration: Codable, Equatable, Sendable {
  let `extension`: String
  let contentType: String
  let bytes: Int64
  let durationSeconds: Double
  let sha256: String
}

enum JobSourceKind: String, Codable, Equatable, Sendable { case url, file }

struct JobSourceMetadata: Codable, Equatable, Sendable {
  let sourceTitle: String?
  let sourceKind: JobSourceKind?
  let clientStartedAt: Date?
  let sourceURL: String?

  init(
    sourceTitle: String? = nil, sourceKind: JobSourceKind? = nil,
    clientStartedAt: Date? = nil, sourceURL: String? = nil
  ) {
    self.sourceTitle = sourceTitle
    self.sourceKind = sourceKind
    self.clientStartedAt = clientStartedAt
    self.sourceURL = sourceURL
  }
}

struct JobTiming: Codable, Equatable, Sendable {
  let processingElapsedMs: Int64?
  let processingElapsedApproximate: Bool
  let totalElapsedMs: Int64?
  let totalElapsedApproximate: Bool
}

struct JobStages: Codable, Equatable, Sendable {
  let validatingAt: Date?
  let processingStartedAt: Date?
  let processingFinishedAt: Date?
  let uploadingResultAt: Date?
}

struct Job: Codable, Equatable, Identifiable, Sendable {
  struct Input: Codable, Equatable, Sendable {
    let `extension`: String
    let bytes: Int64
    let durationSeconds: Double
  }
  struct Failure: Codable, Equatable, Sendable {
    let code: String
    let at: Date
  }
  let id: String
  // An unknown server state is retained so consumers can render it read-only.
  let status: String
  let createdAt: Date
  let updatedAt: Date
  let queuedAt: Date?
  let finishedAt: Date?
  let retryOfJobId: String?
  let input: Input
  let error: Failure?
  let canDownloadInput: Bool
  let canDownloadOutput: Bool
  let workerAvailable: Bool?
  let requestId: String?
  let sourceTitle: String?
  let displayName: String?
  let sourceKind: JobSourceKind?
  let serverTime: Date?
  let timing: JobTiming?
  let stages: JobStages?

  init(
    id: String, status: String, createdAt: Date, updatedAt: Date, queuedAt: Date?,
    finishedAt: Date?, retryOfJobId: String?, input: Input, error: Failure?,
    canDownloadInput: Bool, canDownloadOutput: Bool, workerAvailable: Bool?,
    requestId: String? = nil, sourceTitle: String? = nil, displayName: String? = nil,
    sourceKind: JobSourceKind? = nil, serverTime: Date? = nil, timing: JobTiming? = nil,
    stages: JobStages? = nil
  ) {
    self.id = id
    self.status = status
    self.createdAt = createdAt
    self.updatedAt = updatedAt
    self.queuedAt = queuedAt
    self.finishedAt = finishedAt
    self.retryOfJobId = retryOfJobId
    self.input = input
    self.error = error
    self.canDownloadInput = canDownloadInput
    self.canDownloadOutput = canDownloadOutput
    self.workerAvailable = workerAvailable
    self.requestId = requestId
    self.sourceTitle = sourceTitle
    self.displayName = displayName
    self.sourceKind = sourceKind
    self.serverTime = serverTime
    self.timing = timing
    self.stages = stages
  }

  var preferredName: String {
    let value = displayName ?? sourceTitle
    return value?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
      ? value! : String(localized: "processing_untitled")
  }
}

struct JobPage: Codable, Equatable, Sendable {
  let items: [Job]
  let nextCursor: String?
}

enum UploadMethod: String, Codable, Equatable, Sendable { case put = "PUT" }

struct UploadGrant: Codable, Equatable, Sendable {
  let method: UploadMethod
  let url: URL
  let headers: [String: String]
  let expiresAt: Date
}

struct DownloadGrant: Codable, Equatable, Sendable {
  let url: URL
  let expiresAt: Date
}

struct CreateReservation: Codable, Equatable, Sendable {
  let id: String
  let status: String
  let upload: UploadGrant?
  let requestId: String?

  init(id: String, status: String, upload: UploadGrant?, requestId: String? = nil) {
    self.id = id
    self.status = status
    self.upload = upload
    self.requestId = requestId
  }
}

struct JobMutation: Codable, Equatable, Sendable {
  let id: String
  let status: String
  let retryOfJobId: String?
}

enum JobsFailure: Error, Equatable, Sendable {
  case invalidInput
  case installationRequired
  case forbidden(code: String?)
  case notFound
  case conflict(code: String?)
  case rateLimited(retryAfter: TimeInterval)
  case serviceUnavailable
  case redirectRejected
  case malformedResponse
}

enum ClientErrorStage: String, Codable, CaseIterable, Sendable {
  case unknown = "UNKNOWN"
  case sourceIntake = "SOURCE_INTAKE"
  case downloadingSource = "DOWNLOADING_SOURCE"
  case preparingInput = "PREPARING_INPUT"
  case reservingJob = "RESERVING_JOB"
  case uploadingInput = "UPLOADING_INPUT"
  case confirmingUpload = "CONFIRMING_UPLOAD"
  case refreshingJob = "REFRESHING_JOB"
  case cancelling = "CANCELLING"
  case retrying = "RETRYING"
  case fetchingOutput = "FETCHING_OUTPUT"
  case playback = "PLAYBACK"
  case exporting = "EXPORTING"
}

enum ClientErrorCode: String, Codable, CaseIterable, Sendable {
  case unknown = "UNKNOWN"
  case network = "NETWORK"
  case timeout = "TIMEOUT"
  case authentication = "AUTHENTICATION"
  case invalidMedia = "INVALID_MEDIA"
  case sourceUnavailable = "SOURCE_UNAVAILABLE"
  case storage = "STORAGE"
  case server = "SERVER"
  case jobNotFound = "JOB_NOT_FOUND"
  case jobConflict = "JOB_CONFLICT"
  case checksumMismatch = "CHECKSUM_MISMATCH"
  case localIO = "LOCAL_IO"
}

enum ClientPlatform: String, Codable, Sendable { case ios, android }

struct ClientErrorEvent: Codable, Equatable, Identifiable, Sendable {
  var id: UUID { eventId }
  let eventId: UUID
  let operationId: UUID
  let jobId: String?
  let stage: ClientErrorStage
  let code: ClientErrorCode
  let retryable: Bool
  let platform: ClientPlatform
  let appVersion: String
  let osVersion: String
  let occurredAt: Date
  let httpStatus: Int?
}

struct ClientErrorReceipt: Codable, Equatable, Sendable { let eventId: String }

File: android/app/src/main/java/com/hatem/musicmute/processing/JobModels.kt
package com.hatem.musicmute.processing

import java.time.Instant
import java.time.temporal.ChronoUnit
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerialName
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

@Serializable
data class InputDeclaration(
    val extension: String,
    val contentType: String,
    val bytes: Long,
    val durationSeconds: Double,
    val sha256: String,
)

@Serializable
enum class SourceKind(val wireValue: String) {
    @SerialName("url") URL("url"),
    @SerialName("file") FILE("file"),
}

data class CreateJobMetadata(
    val sourceTitle: String? = null,
    val sourceKind: SourceKind? = null,
    val clientStartedAt: Instant? = null,
    val sourceUrl: String? = null,
    val policyVersion: Int? = null,
    val preparationProfileId: String? = null,
    val source: String? = null,
)

/** Unknown server states remain displayable, but have no actionable enum value. */
enum class JobStatus(val wireValue: String) {
    AWAITING_UPLOAD("awaiting_upload"), QUEUED("queued"), VALIDATING("validating"),
    PROCESSING("processing"), UPLOADING_RESULT("uploading_result"), INTERRUPTED("interrupted"),
    CANCEL_REQUESTED("cancel_requested"), READY("ready"), FAILED("failed"), CANCELLED("cancelled");
}

object JobInstantSerializer : KSerializer<Instant> {
    override val descriptor = PrimitiveSerialDescriptor("JobInstant", PrimitiveKind.STRING)
    override fun serialize(encoder: Encoder, value: Instant) = encoder.encodeString(wireInstant(value))
    override fun deserialize(decoder: Decoder): Instant = Instant.parse(decoder.decodeString())
}

fun wireInstant(value: Instant): String = value.truncatedTo(ChronoUnit.MILLIS).toString()

fun validateDisplayName(value: String): String {
    val result = value.trim()
    require(result.codePointCount(0, result.length) in 1..200)
    require(!result.codePoints().anyMatch(Character::isISOControl))
    return result
}

fun boundedSourceTitle(value: String): String {
    val clean = buildString {
        value.trim().codePoints().forEach { point ->
            appendCodePoint(if (Character.isISOControl(point)) ' '.code else point)
        }
    }.trim().ifBlank { "Audio" }
    val count = clean.codePointCount(0, clean.length)
    return if (count <= 200) clean else clean.substring(0, clean.offsetByCodePoints(0, 200)).trim()
}

@Serializable
data class JobInput(val extension: String, val bytes: Long, val durationSeconds: Double)

@Serializable
data class JobError(
    val code: String,
    val message: String,
    @Serializable(with = JobInstantSerializer::class) val at: Instant,
)

@Serializable
data class JobTiming(
    val processingElapsedMs: Long? = null,
    val processingElapsedApproximate: Boolean = false,
    val totalElapsedMs: Long? = null,
    val totalElapsedApproximate: Boolean = false,
)

@Serializable
data class JobStages(
    @Serializable(with = JobInstantSerializer::class) val validatingAt: Instant? = null,
    @Serializable(with = JobInstantSerializer::class) val processingStartedAt: Instant? = null,
    @Serializable(with = JobInstantSerializer::class) val processingFinishedAt: Instant? = null,
    @Serializable(with = JobInstantSerializer::class) val uploadingResultAt: Instant? = null,
)

@Serializable
data class ServerStageMeasurement(val stage: String, val durationMs: Long, val complete: Boolean)

@Serializable
data class ServerStageTimings(
    val totalMs: Long? = null,
    val totalComplete: Boolean = false,
    val stages: List<ServerStageMeasurement> = emptyList(),
)

@Serializable
data class Job(
    val id: String,
    val status: String,
    @Serializable(with = JobInstantSerializer::class) val createdAt: Instant,
    @Serializable(with = JobInstantSerializer::class) val updatedAt: Instant,
    val input: JobInput,
    val canDownloadInput: Boolean,
    val canDownloadOutput: Boolean,
    @Serializable(with = JobInstantSerializer::class) val queuedAt: Instant? = null,
    @Serializable(with = JobInstantSerializer::class) val finishedAt: Instant? = null,
    val retryOfJobId: String? = null,
    val error: JobError? = null,
    val workerAvailable: Boolean? = null,
    val requestId: String? = null,
    val sourceTitle: String? = null,
    val displayName: String? = null,
    val sourceKind: String? = null,
    @Serializable(with = JobInstantSerializer::class) val serverTime: Instant? = null,
    val timing: JobTiming? = null,
    val serverStageTimings: ServerStageTimings? = null,
    val stages: JobStages? = null,
) {
    val knownStatus: JobStatus? get() = JobStatus.entries.find { it.wireValue == status }
}

@Serializable data class JobPage(val items: List<Job>, val nextCursor: String? = null)

@Serializable
enum class UploadMethod { @SerialName("PUT") PUT }

@Serializable
data class UploadGrant(
    val method: UploadMethod,
    val url: String,
    val headers: Map<String, String>,
    @Serializable(with = JobInstantSerializer::class) val expiresAt: Instant,
)

@Serializable
data class DownloadGrant(
    val url: String,
    @Serializable(with = JobInstantSerializer::class) val expiresAt: Instant,
)

@Serializable data class JobMutation(val id: String, val status: String, val retryOfJobId: String? = null)
@Serializable data class CreateReservation(
    val id: String,
    val status: String,
    val upload: UploadGrant? = null,
    val requestId: String? = null,
)

@Serializable
enum class ClientErrorStage {
    UNKNOWN, SOURCE_INTAKE, DOWNLOADING_SOURCE, PREPARING_INPUT, RESERVING_JOB,
    UPLOADING_INPUT, CONFIRMING_UPLOAD, REFRESHING_JOB, CANCELLING, RETRYING,
    FETCHING_OUTPUT, PLAYBACK, EXPORTING,
}

@Serializable
enum class ClientErrorCode {
    UNKNOWN, NETWORK, TIMEOUT, AUTHENTICATION, INVALID_MEDIA, SOURCE_UNAVAILABLE,
    STORAGE, SERVER, JOB_NOT_FOUND, JOB_CONFLICT, CHECKSUM_MISMATCH, LOCAL_IO,
}

@Serializable
data class ClientErrorReport(
    val eventId: String,
    val operationId: String,
    val jobId: String? = null,
    val stage: ClientErrorStage,
    val code: ClientErrorCode,
    val retryable: Boolean,
    val platform: String,
    val appVersion: String,
    val osVersion: String,
    @Serializable(with = JobInstantSerializer::class) val occurredAt: Instant,
    val httpStatus: Int? = null,
)

@Serializable data class ClientErrorAccepted(val eventId: String)

enum class JobsProblem {
    MEDIA_TOO_LONG, MEDIA_TOO_LARGE, MEDIA_NO_AUDIO, MEDIA_DEFAULT_TRACK_UNAVAILABLE, MEDIA_UNSUPPORTED, MEDIA_DURATION_UNKNOWN,
    YOUTUBE_PLAYLIST_UNSUPPORTED, YOUTUBE_LIVE_UNSUPPORTED, PROCESSING_ALLOWANCE_EXHAUSTED, PROCESSING_QUEUE_FULL,
    PROCESSING_POLICY_INCOMPATIBLE, PROCESSING_CAPACITY_UNAVAILABLE, PROCESSING_LIMIT_REACHED,
    INVALID_INPUT, UNAUTHENTICATED, ACCOUNT_DISABLED, POLICY_DENIED,
    EMAIL_VERIFICATION_REQUIRED, APP_UPDATE_REQUIRED, DEVICE_SYNC_REQUIRED, DEVICE_REPORT_CONFLICT,
    PROFILE_SYNC_REQUIRED, JOB_NOT_FOUND, JOB_STATE_CONFLICT, IDEMPOTENCY_CONFLICT,
    JOB_ACTIVE, UPLOAD_NOT_READY, UPLOAD_RESERVATION_EXPIRED, UPLOAD_GRANT_LIMIT_REACHED,
    UPLOAD_BYTE_LIMIT_REACHED, UPLOAD_ATTEMPT_LIMIT_REACHED, NEW_INPUT_REQUIRED,
    RETAINED_STORAGE_LIMIT_REACHED, DOWNLOAD_RESERVATION_EXPIRED,
    DOWNLOAD_GRANT_LIMIT_REACHED, DOWNLOAD_BYTE_LIMIT_REACHED, SERVICE_BANDWIDTH_LIMIT_REACHED,
    RATE_LIMITED, OFFLINE, SERVICE_UNAVAILABLE,
}

/** Only allowlisted codes are exposed; server diagnostics never become exception text. */
class JobsFailure(val problem: JobsProblem, val retryAfterSeconds: Long? = null) : Exception(problem.name)
